src/shared/vscode/commands2.ts

```diff
@@ -0,0 +1,330 @@
+     * may change in the future.
+     *
+     * #### Example
+     * ```ts
+     * class Foo {
+     *     public square(n: number): number {
+     *         return n * n
+     *     }
+     * }
+     *
+     * const square = Commands.instance.from(Foo).declareSquare('aws.square')
+     * await square.register(new Foo()).execute(5)
+     * ```
+     */
+    public from<T>(target: new (...args: any[]) => T): Declarables<T> {
+        const result = {} as Record<string, (id: string) => DeclaredCommand>
+
+        for (const [k, v] of Object.entries<Callback>(getFunctions(target))) {
+            const mappedKey = `declare${capitalize(k)}`
+            const name = !isNameMangled() ? `${target.name}.${k}` : ''
+
+            result[mappedKey] = id => this.declare({ id, name }, (instance: T) => v.bind(instance))
+        }
+
+        return result as Declarables<T>
+    }
+
+    public dispose(): void {
+        vscode.Disposable.from(...this.resources.values()).dispose()
+    }
+
+    private addResource<T extends Callback, U extends any[]>(resource: CommandResource<T, U>): CommandResource<T, U> {
+        const registered = this.resources.get(resource.id)
+
+        if (registered?.declared) {
+            throw new Error(`Command "${resource.id}" has already been declared by the Toolkit`)
+        }
+
+        this.resources.set(resource.id, resource)
+        return resource
+    }
+
+    static #instance: Commands
+
+    public static get instance(): Commands {
+        return (this.#instance ??= new Commands())
+    }
+}
+
+/**
+ * Short-hand for {@link Commands.get}.
+ */
+export const getCommand = Commands.instance.get.bind(Commands.instance)
+
+/**
+ * Short-hand for {@link Commands.register}.
+ */
+export const registerCommand = Commands.instance.register.bind(Commands.instance)
+
+/**
+ * Short-hand for {@link Commands.declare}.
+ */
+export const declareCommand = Commands.instance.declare.bind(Commands.instance)
+
+/**
+ * Short-hand for {@link Commands.from}.
+ */
+export const fromClass = Commands.instance.from.bind(Commands.instance)
+
+type Functions<T> = { [P in keyof T]: T[P] extends Callback ? T[P] : never }
+type FunctionKeys<T> = { [P in keyof T]: T[P] extends Callback ? P : never }[keyof T]
+
+interface Declare<T, F extends Callback> {
+    (id: string): DeclaredCommand<F, [target: T]>
+}
+
+type Declarables<T> = {
+    [P in FunctionKeys<T> as `declare${Capitalize<P & string>}`]: Declare<T, Functions<T>[P]>
+}
+
+/**
+ * Returns all functions found on the target's prototype chain.
+ *
+ * Conflicts from functions sharing the same key are resolved by order of appearance, earlier
+ * functions given precedence. This is equivalent to how the prototype chain is traversed when
+ * evaluating `target[key]`, so long as the property descriptor is not a 'getter' function.
+ */
+function getFunctions<T>(target: new (...args: any[]) => T): Functions<T> {
```

Comment:
tsUtil / functionUtil?

Reply:
Yeah could move it. It's very niche though. 
This is like a `classUtil` type of function.